File: src/skillguard/engines/structural/schema_validator.py
"""Schema validation engine for skill structure.

Validates that skill files conform to expected structure and
identifies anomalies that could indicate tampering or malicious intent.
"""

from __future__ import annotations

import re
import time

from skillguard.core.models import (
    DetectionRule,
    EngineResult,
    EngineVerdict,
    FileType,
    Finding,
    Severity,
    SkillFile,
)
from skillguard.engines.base import ScanEngine

# Expected sections in a well-formed SKILL.md
_EXPECTED_SECTIONS = {"description", "usage", "instructions", "examples"}

# Maximum reasonable file sizes (bytes)
_MAX_SKILL_MD_SIZE = 100_000
_MAX_SCRIPT_SIZE = 500_000


class SchemaValidator(ScanEngine):
    """Validates skill structure against expected schemas."""

    @property
    def name(self) -> str:
        return "schema_validator"

    @property
    def version(self) -> str:
        return "0.3.0"

    async def scan(
        self,
        skill_files: list[SkillFile],
        rules: list[DetectionRule] | None = None,
    ) -> EngineResult:
        start = time.monotonic()
        findings: list[Finding] = []

        # Check overall skill structure
        findings.extend(self._validate_structure(skill_files))

        # Validate individual files
        for sf in skill_files:
            findings.extend(self._validate_file(sf))
            if sf.file_type in {FileType.SCRIPT_PYTHON, FileType.SCRIPT_BASH, FileType.SCRIPT_JS}:
                findings.extend(self._check_deserialization_safety(sf))

        elapsed_ms = int((time.monotonic() - start) * 1000)

        if not findings:
            return EngineResult(
                engine_name=self.name,
                engine_version=self.version,
                verdict=EngineVerdict.CLEAN,
                confidence=0.8,
                findings=[],
                duration_ms=elapsed_ms,
            )

        max_confidence = max(f.confidence for f in findings)
        severities = {f.severity for f in findings}
        if Severity.CRITICAL in severities or Severity.HIGH in severities:
            verdict = EngineVerdict.SUSPICIOUS
        elif Severity.MEDIUM in severities:
            verdict = EngineVerdict.SUSPICIOUS
        else:
            verdict = EngineVerdict.CLEAN

        return EngineResult(
            engine_name=self.name,
            engine_version=self.version,
            verdict=verdict,
            confidence=max_confidence,
            detection_name="Schema Validation" if findings else None,
            findings=findings,
            duration_ms=elapsed_ms,
        )

    async def health_check(self) -> bool:
        return True

    def _validate_structure(self, skill_files: list[SkillFile]) -> list[Finding]:
        findings: list[Finding] = []

        # Check for SKILL.md presence
        has_skill_md = any(sf.file_type == FileType.SKILL_MD for sf in skill_files)
        if not has_skill_md:
            findings.append(
                Finding(
                    rule_id="SG-STRUCT-001",
                    rule_name="Missing SKILL.md",
                    severity=Severity.LOW,
                    category="structural",
                    description="No SKILL.md found. Skills should have a main description file.",
                    file_path="(skill root)",
                    confidence=0.90,
                )
            )

        # Check for suspiciously many executable scripts
        script_count = sum(
            1 for sf in skill_files
            if sf.file_type in {FileType.SCRIPT_PYTHON, FileType.SCRIPT_BASH, FileType.SCRIPT_JS}
        )
        if script_count > 10:
            findings.append(
                Finding(
                    rule_id="SG-STRUCT-002",
                    rule_name="Excessive Script Files",
                    severity=Severity.MEDIUM,
                    category="structural",
                    description=f"Skill contains {script_count} executable scripts, which is unusually high.",
                    file_path="(skill root)",
                    confidence=0.65,
                    remediation="Review whether all scripts are necessary.",
                )
            )

        # Check for hidden files
        hidden_files = [sf for sf in skill_files if sf.path.split("/")[-1].startswith(".")]
        for sf in hidden_files:
            findings.append(
                Finding(
                    rule_id="SG-STRUCT-003",
                    rule_name="Hidden File Detected",
                    severity=Severity.LOW,
                    category="structural",
                    description=f"Hidden file '{sf.path}' detected in skill package.",
                    file_path=sf.path,
                    confidence=0.60,
                )
            )

        return findings

    def _check_deserialization_safety(self, sf: SkillFile) -> list[Finding]:
        """Check for unsafe deserialization patterns (OWASP AST05)."""
        findings: list[Finding] = []
        content = sf.content or ""

        # SG-STRUCT-DS-001: Unsafe XML parsing
        xml_pattern = re.compile(
            r"(?i)(xml\.etree\.ElementTree|xml\.dom\.minidom|xml\.sax|lxml\.etree)(?!.*defusedxml)"
        )
        for match in xml_pattern.finditer(content):
            line_num = content[:match.start()].count("\n") + 1
            findings.append(
                Finding(
                    rule_id="SG-STRUCT-DS-001",
                    rule_name="Unsafe XML Parsing",
                    severity=Severity.HIGH,
                    category="structural",
                    description=(
                        f"Unsafe XML parser '{match.group(1)}' used without defusedxml. "
                        f"Vulnerable to XML external entity (XXE) attacks."
                    ),
                    file_path=sf.path,
                    line_start=line_num,
                    snippet=match.group()[:200],
                    owasp_ast=["AST05"],
                    confidence=0.85,
                    remediation="Use defusedxml instead of standard XML parsers to prevent XXE attacks.",
                )
            )

        # SG-STRUCT-DS-002: JSON with code execution hooks
        json_hook_pattern = re.compile(
            r"(?i)json\.loads?\s*\(.*object_hook|jsonpickle\.decode"
        )
        for match in json_hook_pattern.finditer(content):
            line_num = content[:match.start()].count("\n") + 1
            findings.append(
                Finding(
                    rule_id="SG-STRUCT-DS-002",
                    rule_name="JSON Code Execution Hook",
                    severity=Severity.HIGH,
                    category="structural",
                    description=(
                        "JSON deserialization with code execution hooks detected. "
                        "This can lead to arbitrary code execution via crafted payloads."
                    ),
                    file_path=sf.path,
                    line_start=line_num,
                    snippet=match.group()[:200],
                    owasp_ast=["AST05"],
                    confidence=0.80,
                    remediation="Avoid using object_hook with untrusted data. Do not use jsonpickle for untrusted input.",
                )
            )

        return findings

    def _validate_file(self, sf: SkillFile) -> list[Finding]:
        findings: list[Finding] = []

        # Check file size limits
        if sf.file_type == FileType.SKILL_MD and sf.size_bytes > _MAX_SKILL_MD_SIZE:
            findings.append(
                Finding(
                    rule_id="SG-STRUCT-004",
                    rule_name="Oversized SKILL.md",
                    severity=Severity.MEDIUM,
                    category="structural",
                    description=f"SKILL.md is {sf.size_bytes:,} bytes, exceeding the {_MAX_SKILL_MD_SIZE:,} byte limit.",
                    file_path=sf.path,
                    confidence=0.75,
                    remediation="Reduce SKILL.md size. Large files may contain hidden payloads.",
                )
            )

        # Check for binary content in text files
        if sf.content and sf.file_type in {FileType.SKILL_MD, FileType.FRONTMATTER}:
            null_count = sf.content.count("\x00")
            if null_count > 0:
                findings.append(
                    Finding(
                        rule_id="SG-STRUCT-005",
                        rule_name="Binary Content in Text File",
                        severity=Severity.HIGH,
                        category="structural",
                        description="Text file contains null bytes, suggesting binary content or data hiding.",
                        file_path=sf.path,
                        confidence=0.85,
                    )
                )

        return findings

File: go-skillguard/internal/engines/schema_validator.go
package engines

import (
	"encoding/json"
	"fmt"
	"regexp"
	"strings"
	"time"

	"github.com/girdav01/skillguard/internal/core"
)

// SchemaValidator validates MCP tool schemas and skill structure.
type SchemaValidator struct{}

func NewSchemaValidator() *SchemaValidator { return &SchemaValidator{} }

func (s *SchemaValidator) Name() string    { return "schema_validator" }
func (s *SchemaValidator) Version() string { return "0.3.0" }

func (s *SchemaValidator) Scan(skillFiles []core.SkillFile, rules []core.DetectionRule) (*core.EngineResult, error) {
	start := time.Now()
	var findings []core.Finding

	hasSkillMD := false
	for _, sf := range skillFiles {
		if sf.FileType == core.FileTypeSkillMD {
			hasSkillMD = true
		}
		if sf.Content == nil {
			continue
		}
		if sf.FileType == core.FileTypeConfig || strings.HasSuffix(sf.Path, ".json") {
			findings = append(findings, s.validateJSON(sf)...)
		}
		if sf.FileType == core.FileTypeSkillMD {
			findings = append(findings, s.validateSkillMD(sf)...)
		}
	}

	// Check deserialization safety for script files
	for _, sf := range skillFiles {
		if sf.Content == nil {
			continue
		}
		if sf.FileType == core.FileTypeScriptPython || sf.FileType == core.FileTypeScriptBash || sf.FileType == core.FileTypeScriptJS || sf.FileType == core.FileTypeScriptTS {
			findings = append(findings, s.checkDeserializationSafety(sf)...)
		}
	}

	if !hasSkillMD && len(skillFiles) > 0 {
		remediation := "Add a SKILL.md file with proper metadata and description."
		findings = append(findings, core.Finding{
			RuleID:      "SG-SCHEMA-001",
			RuleName:    "Missing SKILL.md",
			Severity:    core.SeverityLow,
			Category:    "schema",
			Description: "Skill package does not contain a SKILL.md file with metadata.",
			FilePath:    ".",
			Confidence:  0.95,
			Remediation: &remediation,
		})
	}

	elapsed := time.Since(start).Milliseconds()

	if len(findings) == 0 {
		return &core.EngineResult{
			EngineName:    s.Name(),
			EngineVersion: s.Version(),
			Verdict:       core.EngineVerdictClean,
			Confidence:    0.9,
			Findings:      []core.Finding{},
			DurationMs:    elapsed,
		}, nil
	}

	verdict := computeVerdict(findings)
	return &core.EngineResult{
		EngineName:    s.Name(),
		EngineVersion: s.Version(),
		Verdict:       verdict,
		Confidence:    computeConfidence(findings),
		Findings:      findings,
		DurationMs:    elapsed,
	}, nil
}

func (s *SchemaValidator) HealthCheck() bool { return true }

func (s *SchemaValidator) validateJSON(sf core.SkillFile) []core.Finding {
	var findings []core.Finding
	var data any
	if err := json.Unmarshal([]byte(*sf.Content), &data); err != nil {
		lineNum := 1
		snippet := err.Error()
		remediation := "Fix the JSON syntax error."
		findings = append(findings, core.Finding{
			RuleID:      "SG-SCHEMA-002",
			RuleName:    "Invalid JSON",
			Severity:    core.SeverityMedium,
			Category:    "schema",
			Description: fmt.Sprintf("JSON file '%s' has syntax errors.", sf.Path),
			FilePath:    sf.Path,
			LineStart:   &lineNum,
			Snippet:     &snippet,
			Confidence:  0.95,
			Remediation: &remediation,
		})
	}
	return findings
}

func (s *SchemaValidator) validateSkillMD(sf core.SkillFile) []core.Finding {
	var findings []core.Finding
	content := *sf.Content

	fm, _ := core.ParseFrontmatter(content)
	if len(fm) == 0 {
		remediation := "Add YAML frontmatter with metadata (name, version, description)."
		findings = append(findings, core.Finding{
			RuleID:      "SG-SCHEMA-003",
			RuleName:    "Missing Frontmatter",
			Severity:    core.SeverityLow,
			Category:    "schema",
			Description: "SKILL.md file is missing YAML frontmatter with metadata.",
			FilePath:    sf.Path,
			Confidence:  0.90,
			Remediation: &remediation,
		})
	}
	return findings
}

func (s *SchemaValidator) checkDeserializationSafety(sf core.SkillFile) []core.Finding {
	var findings []core.Finding
	content := *sf.Content

	// SG-STRUCT-DS-001: Unsafe XML parsing
	xmlPattern := regexp.MustCompile(`(?i)(xml\.etree\.ElementTree|xml\.dom\.minidom|xml\.sax|lxml\.etree)`)
	defusedPattern := regexp.MustCompile(`(?i)defusedxml`)
	for _, match := range xmlPattern.FindAllStringIndex(content, -1) {
		// Check if defusedxml appears on the same line
		lineStart := strings.LastIndex(content[:match[0]], "\n") + 1
		lineEnd := strings.Index(content[match[1]:], "\n")
		if lineEnd < 0 {
			lineEnd = len(content)
		} else {
			lineEnd += match[1]
		}
		line := content[lineStart:lineEnd]
		if defusedPattern.MatchString(line) {
			continue
		}
		lineNum := strings.Count(content[:match[0]], "\n") + 1
		snippet := content[match[0]:match[1]]
		remediation := "Use defusedxml instead of standard XML parsers to prevent XXE attacks."
		findings = append(findings, core.Finding{
			RuleID:      "SG-STRUCT-DS-001",
			RuleName:    "Unsafe XML Parsing",
			Severity:    core.SeverityHigh,
			Category:    "structural",
			Description: fmt.Sprintf("Unsafe XML parser '%s' used without defusedxml. Vulnerable to XML external entity (XXE) attacks.", snippet),
			FilePath:    sf.Path,
			LineStart:   &lineNum,
			Snippet:     &snippet,
			OWASPAST:    []string{"AST05"},
			Confidence:  0.85,
			Remediation: &remediation,
		})
	}

	// SG-STRUCT-DS-002: JSON with code execution hooks
	jsonHookPattern := regexp.MustCompile(`(?i)json\.loads?\s*\(.*object_hook|jsonpickle\.decode`)
	for _, match := range jsonHookPattern.FindAllStringIndex(content, -1) {
		lineNum := strings.Count(content[:match[0]], "\n") + 1
		snippet := content[match[0]:match[1]]
		if len(snippet) > 200 {
			snippet = snippet[:200]
		}
		remediation := "Avoid using object_hook with untrusted data. Do not use jsonpickle for untrusted input."
		findings = append(findings, core.Finding{
			RuleID:      "SG-STRUCT-DS-002",
			RuleName:    "JSON Code Execution Hook",
			Severity:    core.SeverityHigh,
			Category:    "structural",
			Description: "JSON deserialization with code execution hooks detected. This can lead to arbitrary code execution via crafted payloads.",
			FilePath:    sf.Path,
			LineStart:   &lineNum,
			Snippet:     &snippet,
			OWASPAST:    []string{"AST05"},
			Confidence:  0.80,
			Remediation: &remediation,
		})
	}

	return findings
}
